beacon_chain/state/active_state.py

```diff
@@ -1,44 +1,31 @@
-from .partial_crosslink_record import PartialCrosslinkRecord
-from .recent_proposer_record import RecentProposerRecord
+from .attestation_record import AttestationRecord
 
 
 class ActiveState():
 
     fields = {
-        # Block height
-        'height': 'int64',
-        # Global RANDAO beacon state
-        'randao': 'hash32',
-        # Which validators have made FFG votes this epoch (as a bitfield)
-        'ffg_voter_bitfield': 'bytes',
-        # Block attesters in the last epoch
-        'recent_attesters': ['int24'],
-        # Storing data about crosslinks-in-progress attempted in this epoch
-        'partial_crosslinks': [PartialCrosslinkRecord],
-        # Total number of skips (used to determine minimum timestamp)
-        'total_skip_count': 'int64',
-        # Block proposers in the last epoch
-        'recent_proposers': [RecentProposerRecord]
+        # Attestations that have not yet been processed
+        'pending_attestations': [AttestationRecord],
+        # Most recent 2*CYCLE_LENGTH block hashes, older to newer
+        'recent_block_hashes': ['hash32'],
     }
     defaults = {
-        'height': 0,
-        'randao': b'\x00'*32,
-        'ffg_voter_bitfield': b'',
-        'recent_attesters': [],
-        'partial_crosslinks': [],
-        'total_skip_count': 0,
-        'recent_proposers': []
+        'pending_attestations': [],
+        'recent_block_hashes': [],
     }
 
     def __init__(self, **kwargs):
         for k in self.fields.keys():
             assert k in kwargs or k in self.defaults
             setattr(self, k, kwargs.get(k, self.defaults.get(k)))
 
-    @property
-    def num_recent_attesters(self):
-        return len(self.recent_attesters)
+        # block_vote_cache is not part of protocol state
+        # is used as a helper cache to aid in doing the cycle init calculations
```

Comment:
1. It's not the usual disk cache or CPU cache, right? What's the lifecycle of `block_vote_cache`?
2. What do you think about implementing it with function caching and decouple with `ActiveState` once we integrated with DB (with py-evm) ? e.g., assume that the client has to keep recent `2 * CYCLE_LENGTH` states and related votes info.

Reply:
1. Lifecycle is just as long as we are keeping around a particular `ActiveState` instance. It's easy when adding attestations to remember things about the votes. We can walk back in the state and do this during a cycle transition but caching at this time just seemed more obvious to me than the other route. It's not disk/cpu cache. Just a temporary place to store some extra bits.
2. function caching? Yeah, we can easily decouple from living inside the ActiveState model itself. In the current architecture, I didn't see an obvious place to put it without building something new. This was a somewhat lazy choice on my part 😛